### PR TITLE
Incoming webhooks token fix

### PR DESCRIPTION
**Pull request by Joshix-1**
Not every incoming webhook comes with a token. IncomingWebhooks without a token are useless. So I think we should just ignore them. https://github.com/discord/discord-api-docs/issues/3056

### CODE REVIEW

**Comment by coaxial1** [comment]
Hm. Why are they useless? What if I just want to get a list of incoming webhooks in a channel, without posting to them? :)

**Comment by Joshix-1** [comment]
The only real point of incoming webhooks is that they have a token and that you are allowed to post to them.
Having the token Optional would defeat the entire purpose of the IncomingWebhook interface. 
You can still get all Webhooks and then filter out all the webhooks with type Incoming, I know that's stupid, but this pr is the quickest non-breaking change I could make.

We could maybe rename the interface to WebhookWithToken or something.

**Comment by Joshix-1** [comment]
I changed it a bit, to allow getting all real incoming webhooks.
This is the best solution without breaking changes, I can think of right now. 

Things to consider:
- rename functions as the naming is confusing
- document things different
- make things more clear 
- rename IncomingWebhook to somthing with Token or Messageable

**Review by coaxial1 — APPROVED** [review]

File: javacord-api/src/main/java/org/javacord/api/entity/channel/TextChannel.java [hunk 1]
```diff
@@ -841,7 +841,15 @@ default NonThrowingAutoCloseable typeContinuouslyAfter(
     CompletableFuture<List<Webhook>> getWebhooks();
 
     /**
-     * Gets a list of all incoming webhooks in this channel.
+     * Gets a list of all incoming webhooks in this channel, they are not guaranteed to have an accessible token.
+     *
+     * @return A list of all incoming webhooks in this channel.
+     */
+    CompletableFuture<List<Webhook>> getRealIncomingWebhooks();
```
Comment (KILLEliteMaste) [inline javacord-api/src/main/java/org/javacord/api/entity/channel/TextChannel.java:848]: If I see this method I would have no clue what a "Real" incoming webhook would be. If it's just a different type of webhook which may not have a token it should rather be called `getAllIncomingWebhooks`

File: javacord-api/src/main/java/org/javacord/api/entity/server/Server.java [hunk 2]
```diff
@@ -1992,7 +1992,15 @@ default ServerUpdater createUpdater() {
     CompletableFuture<List<Webhook>> getWebhooks();
 
     /**
-     * Gets a list of all incoming webhooks in this server.
+     * Gets a list of all incoming webhooks in this server, they are not guaranteed to have an accessible token.
+     *
+     * @return A list of all incoming webhooks in this server.
+     */
+    CompletableFuture<List<Webhook>> getRealIncomingWebhooks();
```
Comment (KILLEliteMaste) [inline javacord-api/src/main/java/org/javacord/api/entity/server/Server.java:1999]: Same here as above

File: javacord-core/src/main/java/org/javacord/core/entity/channel/InternalTextChannel.java [hunk 3]
```diff
@@ -186,14 +187,30 @@
...
         return new RestRequest<List<IncomingWebhook>>(getApi(), RestMethod.GET, RestEndpoint.CHANNEL_WEBHOOK)
                 .setUrlParameters(getIdAsString())
                 .execute(result -> {
                     List<IncomingWebhook> webhooks = new ArrayList<>();
                     for (JsonNode webhookJson : result.getJsonBody()) {
-                        if (webhookJson.get("type").asText().equals("1")) {
+                        // check if it has a token, as only incoming webhooks have one
+                        if (webhookJson.hasNonNull("token")) {
                             webhooks.add(new IncomingWebhookImpl(getApi(), webhookJson));
                         }
```
Comment (KILLEliteMaste) [inline javacord-core/src/main/java/org/javacord/core/entity/channel/InternalTextChannel.java:215]: Why don't you check here if it's a webhhok of type `INCOMING` like in `getRealIncomingWebhooks`?

File: javacord-core/src/main/java/org/javacord/core/entity/server/ServerImpl.java [hunk 4]
```diff
@@ -1433,14 +1433,30 @@ public void selfUndeafen() {
...
+    }
+
     @Override
     public CompletableFuture<List<IncomingWebhook>> getIncomingWebhooks() {
         return new RestRequest<List<IncomingWebhook>>(getApi(), RestMethod.GET, RestEndpoint.SERVER_WEBHOOK)
                 .setUrlParameters(getIdAsString())
                 .execute(result -> {
                     List<IncomingWebhook> webhooks = new ArrayList<>();
                     for (JsonNode webhookJson : result.getJsonBody()) {
-                        if (WebhookType.fromValue(webhookJson.get("type").asInt()) == WebhookType.INCOMING) {
```
Comment (KILLEliteMaste) [inline javacord-core/src/main/java/org/javacord/core/entity/server/ServerImpl.java:1443]: Same here. Why did you remove the check?

File: javacord-core/src/main/java/org/javacord/core/entity/channel/InternalTextChannel.java (hunk 3 above)
Comment (Joshix-1) [reply javacord-core/src/main/java/org/javacord/core/entity/channel/InternalTextChannel.java:215]: I didn't think the check was necessary.

Comment (KILLEliteMaste) [reply javacord-core/src/main/java/org/javacord/core/entity/channel/InternalTextChannel.java:215]: But then why did you leave it in the newly added `getRealIncomingWebhooks` method?

Comment (Joshix-1) [reply javacord-core/src/main/java/org/javacord/core/entity/channel/InternalTextChannel.java:215]: ups, that was a mistake sorry

Comment (KILLEliteMaste) [reply javacord-core/src/main/java/org/javacord/core/entity/channel/InternalTextChannel.java:215]: Also I would leave the check. Because once discord decides to return a different type of webhook, this may no longer work

Comment (RedsTom) [reply javacord-core/src/main/java/org/javacord/core/entity/channel/InternalTextChannel.java:215]: Maybe combining the two conditions should be fine :
```java
/* 213 */ if (webhookJson.hasNonNull("token") && webhookJson.get("type").asText().equals("1")) {
/* 214 */     webhooks.add(new IncomingWebhookImpl(getApi(), webhookJson));
/* 215 */ }
```

**Comment by Joshix-1** [comment]
I think I fixed all the addressed issues. I rebased some methods to remove duplicated code.

**Review by KILLEliteMaste — APPROVED** [review]
LGTM!
Can you please rebase to the latest version?
